tests/test_client.py

```diff
@@ -108,5 +120,157 @@ async def launch_client() -> None:
+        ),
+    ) as client:
+        async with anyio.create_task_group() as tg:
+            await client.subscribe(topic)
+            tg.start_soon(handle_messages, tg)
+            await client.publish(topic)
+
+
+async def test_client_username_password() -> None:
+    topic = TOPIC_HEADER + "username_password"
+
+    async def handle_messages(tg: anyio.abc.TaskGroup) -> None:
+        async with client.filtered_messages(topic) as messages:
+            async for message in messages:
+                assert message.topic == topic
+                tg.cancel_scope.cancel()
+
+    async with Client(HOSTNAME, username="asyncio-mqtt", password="012") as client:
+        async with anyio.create_task_group() as tg:
+            await client.subscribe(topic)
+            tg.start_soon(handle_messages, tg)
+            await client.publish(topic)
+
+
+async def test_client_logger() -> None:
+    logger = logging.getLogger("asyncio-mqtt")
+    async with Client(HOSTNAME, logger=logger) as client:
+        assert logger == client._client._logger  # type: ignore[attr-defined]
+
+
+async def test_client_max_concurrent_outgoing_calls(
+    monkeypatch: pytest.MonkeyPatch,
+) -> None:
+    topic = TOPIC_HEADER + "max_concurrent_outgoing_calls"
+
+    class MockPahoClient(mqtt.Client):
+        def subscribe(
+            self,
+            topic: str
+            | tuple[str, mqtt.SubscribeOptions]
+            | list[tuple[str, mqtt.SubscribeOptions]]
+            | list[tuple[str, int]],
+            qos: int = 0,
+            options: mqtt.SubscribeOptions | None = None,
+            properties: mqtt.Properties | None = None,
+        ) -> tuple[int, int]:
+            assert client._outgoing_calls_sem is not None
+            assert client._outgoing_calls_sem.locked()
+            return super().subscribe(topic, qos, options, properties)
+
+        def unsubscribe(
+            self, topic: str | list[str], properties: mqtt.Properties | None = None
+        ) -> tuple[int, int]:
+            assert client._outgoing_calls_sem is not None
+            assert client._outgoing_calls_sem.locked()
+            return super().unsubscribe(topic, properties)
+
+        def publish(
+            self,
+            topic: str,
+            payload: PayloadType | None = None,
+            qos: int = 0,
+            retain: bool = False,
+            properties: mqtt.Properties | None = None,
+        ) -> mqtt.MQTTMessageInfo:
+            assert client._outgoing_calls_sem is not None
+            assert client._outgoing_calls_sem.locked()
+            return super().publish(topic, payload, qos, retain, properties)
+
+    monkeypatch.setattr(mqtt, "Client", MockPahoClient)
+
+    async with Client(HOSTNAME, max_concurrent_outgoing_calls=1) as client:
+        await client.subscribe(topic)
+        await client.unsubscribe(topic)
+        await client.publish(topic)
+
+
+async def test_client_websockets() -> None:
```

Comment:
Really good to cover ground here. 👍 I know that many of our users use MQTT piped through a WebSocket connection.